### PR TITLE
Translate budget and budget phase main link url

### PR DESCRIPTION
**Pull request by Senen**
## References

Related PR's: 
 * Improve budget header #4501
 * Improve budget phases #4510 

## Objectives

Make the `main_link_url` field translatable to store different destinations for each language, particularly useful when the destination page is available in many languages.

## Visual Changes

**Before: Budget edition page**
Captura de pantalla 2021-11-03 a las 16 20 39-fullpage

**Before: Budget phase edition page**
Captura de pantalla 2021-11-03 a las 16 20 49-fullpage

**After: Budget edition page**
Captura de pantalla 2021-11-04 a las 11 18 15-fullpage

**After: Budget phase edition page**
Captura de pantalla 2021-11-04 a las 11 20 47-fullpage

## Notes

I omitted the data migration process from non-translatable fields to the new translatable ones because the related PR's were not released yet.

### CODE REVIEW

File: app/models/budget/phase.rb [hunk 1]
```diff
@@ -19,9 +20,9 @@ class Phase < ApplicationRecord
 
     validates_translation :name, presence: true
     validates_translation :description, length: { maximum: DESCRIPTION_MAX_LENGTH }
+    validates_translation :main_link_url, presence: { if: -> { main_link_text.present? }}
```
Comment (javierm) [inline app/models/budget/phase.rb:23]:
```suggestion
    validates_translation :main_link_url, presence: true, unless: -> { main_link_text.blank? }
```

For the same reasons mentioned in the comment about the budget model.

File: config/locales/en/activerecord.yml [hunk 2]
```diff
@@ -160,9 +160,9 @@ en:
         voting_style: "Final voting style"
```
Comment (javierm) [inline config/locales/en/activerecord.yml:160]: We can apply the changes on this file to `es/activerecord.yml` as well :+1:.

File: db/migrate/20211103090940_translate_budget_main_link_url.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class TranslateBudgetMainLinkUrl < ActiveRecord::Migration[5.2]
+  def change
+    add_column :budget_translations, :main_link_url, :string
+    remove_column :budgets, :main_link_url
```
Comment (javierm) [inline db/migrate/20211103090940_translate_budget_main_link_url.rb:4]:
```suggestion
    remove_column :budgets, :main_link_url, :string
```

This way we make the migration reversible :wink:.

File: db/migrate/20211103112944_translate_budget_phase_main_link_url.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+class TranslateBudgetPhaseMainLinkUrl < ActiveRecord::Migration[5.2]
+  def change
+    add_column :budget_phase_translations, :main_link_url, :string
+    remove_column :budget_phases, :main_link_url
```
Comment (javierm) [inline db/migrate/20211103112944_translate_budget_phase_main_link_url.rb:4]:
```suggestion
    remove_column :budget_phases, :main_link_url, :string
```

This way we make the migration reversible :wink:.

File: spec/models/concerns/globalizable.rb [hunk 5]
```diff
@@ -11,6 +11,15 @@
...
   let(:attribute) { required_fields.sample || fields.sample }
+  let(:invalid_translation_attributes) do
+    Hash.new.tap do |translation_attributes|
+      translation_attributes[:id] = record.translations.find_by(locale: :es).id
+      translation_attributes[attribute] = ""
+      if attribute == :main_link_url && [:budget, :budget_phase].include?(factory_name)
+        translation_attributes[:main_link_text] = "Enlace"
+      end
+    end
+  end
```
Comment (javierm) [inline spec/models/concerns/globalizable.rb:22]: Why are these changes necessary? :thinking:

**Review by javierm — CHANGES_REQUESTED** [review]
Awesome! :tada: I've left some comments; let me know what you think!

File: app/models/budget.rb [hunk 6]
```diff
@@ -24,11 +24,11 @@ def name_uniqueness_by_budget
   VOTING_STYLES = %w[knapsack approval].freeze
 
   validates_translation :name, presence: true
+  validates_translation :main_link_url, presence: { if: -> { main_link_text.present? }}
```
[the remark this answers is not in the document]
Comment (Senen) [reply app/models/budget.rb:27]: Yes, this way to define validator options was weird for me too, but [it seems to be a valid way to define validators] particularly useful if you want to define more than one validator for a field in one line. (https://github.com/rails/rails/blob/v5.2.6/activemodel/lib/active_model/validations/validates.rb#L101)

File: app/components/admin/budgets/form_component.html.erb [hunk 7]
```diff
@@ -21,13 +21,12 @@
...
         </div>
       </div>
-    <% end %>
-
-    <div class="row expanded">
-      <div class="small-12 medium-9 large-6 column end">
-        <%= f.text_field :main_link_url, placeholder: t("admin.shared.example_url") %>
+      <div class="row expanded">
+        <div class="small-12 medium-9 large-6 column end">
+          <%= translations_form.text_field :main_link_url, placeholder: t("admin.shared.example_url") %>
```
[the remark this answers is not in the document]
Comment (Senen) [reply app/components/admin/budgets/form_component.html.erb]: Done!

File: spec/models/concerns/globalizable.rb (hunk 5 above)
Comment (Senen) [reply spec/models/concerns/globalizable.rb:22]: Because `globalizable` specs do not take into account conditioned validations. The `main_link_url` field validation only takes place when the field `main_link_text` is fulfilled. The globalizable spec method that fulfils the required fields does not fulfil the `main_link_text`as it is not mandatory and then some spec crashes when executed for the `Budget` or `Budget::Phase` model. This is the fastest way I found to solve it. Maybe there is a better approach 🤔 .

File: app/components/admin/budget_phases/form_component.html.erb [hunk 8]
```diff
@@ -49,11 +49,11 @@
       <p class="help-text"> <%= t("admin.budget_phases.edit.main_call_to_action_description") %></p>
       <%= translations_form.text_field :main_link_text %>
     </div>
-  <% end %>
 
-  <div class="small-12 column">
-    <%= f.text_field :main_link_url, placeholder: t("admin.shared.example_url") %>
-  </div>
+    <div class="small-12 column">
+      <%= translations_form.text_field :main_link_url, placeholder: t("admin.shared.example_url") %>
```
[the remark this answers is not in the document]
Comment (Senen) [reply app/components/admin/budget_phases/form_component.html.erb]: Done!

File: config/locales/en/activerecord.yml (hunk 2 above)
Comment (Senen) [reply config/locales/en/activerecord.yml:160]: Done!

File: db/migrate/20211103090940_translate_budget_main_link_url.rb (hunk 3 above)
Comment (Senen) [reply db/migrate/20211103090940_translate_budget_main_link_url.rb]: Done!

File: db/migrate/20211103112944_translate_budget_phase_main_link_url.rb (hunk 4 above)
Comment (Senen) [reply db/migrate/20211103112944_translate_budget_phase_main_link_url.rb]: Done!

File: app/models/budget.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (Senen) [reply app/models/budget.rb]: > Awesome! 🎉 I've left some comments; let me know what you think!

Wow! It seems this time I wanted to run too much.

**Comment by Senen** [comment]
> Awesome! 🎉 I've left some comments; let me know what you think!

Wow! It seems this time I wanted to run too much.

File: spec/models/concerns/globalizable.rb (hunk 5 above)
Comment (javierm) [reply spec/models/concerns/globalizable.rb:22]: Two possible solutions come to mind :thinking:.

One would be defining an extra factory, like it was done for other models (see commit 66334b5757).

```ruby
  factory :budget do
  # ...
    factory :budget_with_link do
      main_link_text { "Link" }
      main_link_url { "/" }
    end
  end
 
  # Then, in the budget spec:
  it_behaves_like "globalizable", :budget_with_link
```

The other would be changing the way we mark translations as required in order to handle the case where translatable fields are required but only under certain conditions. Here's the patch:

```diff
diff --git a/app/models/concerns/globalizable.rb b/app/models/concerns/globalizable.rb
index 98613b5a7..4c47249bd 100644
--- a/app/models/concerns/globalizable.rb
+++ b/app/models/concerns/globalizable.rb
@@ -32,14 +32,9 @@ module Globalizable
     private
 
       def required_attribute?(attribute)
-        presence_validators = [ActiveModel::Validations::PresenceValidator,
-          ActiveRecord::Validations::PresenceValidator]
-
-        attribute_validators(attribute).any? { |validator| presence_validators.include? validator }
-      end
-
-      def attribute_validators(attribute)
-        self.class.validators_on(attribute).map(&:class)
+        self.class.validators_on(attribute).any? do |validator|
+          validator.kind == :presence && !validator.options[:unless]
+        end
       end
 
       def check_translations_number
```

And here's the test case for this patch :sweat_smile:. I had to add the `validate :check_translations_number, if: :translations_required?` because I didn't manage to add a test using the `on: :update` condition. Due to this, I'm not even sure we're testing the right scenario here :thinking:.

```ruby
require "rails_helper"

describe Globalizable do
  it "doesn't require a translation when translatable fields require conditional presence" do
    dummy_translation = Class.new do
      include ActiveModel::Model
      attr_accessor :title, :locale
      validates :locale, presence: true
    end

    stub_const("DummyTranslation", dummy_translation)

    dummy_model = Class.new do
      include ActiveModel::Model
      attr_accessor :title, :translations

      def self.globalize_accessors; end
      def self.accepts_nested_attributes_for(...); end
      def self.after_validation(...); end
      def self.paranoid?; end

      def self.translation_class
        DummyTranslation
      end

      def translated_attribute_names
        [:title]
      end

      include Globalizable
      validate :check_translations_number, if: :translations_required?
      validates_translation :title, presence: true, unless: -> { true }

      def translations
        @translations ||= []
      end
    end

    stub_const("DummyModel", dummy_model)

    expect(DummyModel.new).to be_valid
  end
end
```

The advantage of the second solution is we could remove the factories defined in commit 66334b5757. But even so, I'm not sure which one I like better, so up to you!

**Review by javierm — APPROVED** [review]
